flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointSerializer.java

```diff
@@ -52,6 +37,5 @@
 	 * @return The deserialized savepoint
 	 * @throws IOException Serialization failures are forwarded
 	 */
-	T deserialize(DataInputStream dis, ClassLoader userCodeClassLoader) throws IOException;
-
+	SavepointV2 deserialize(DataInputStream dis, ClassLoader userCodeClassLoader) throws IOException;
```

Comment:
Should we bind #SavepointSerializer to #SavepointV2 ? 
In this way, we have to create a new interface if we want to add #SavepointV3 in the future. Is that expected?

Reply:
In this design, this is expected. If we introduce a new format version of the metadata, all deserializers need to deserialize to that new version.

Reply:
This has always been the case in the code. Previously, the interface "pretended" to be cross version generic, but in the end all deserializers had to be updated every time a new version came. This now makes this explicit.

Reply:
You're right. I was misguided by the previous interface.

Reply:
This change here contains something similar to your suggestion: https://issues.apache.org/jira/browse/FLINK-16259

Reply:
The related PR is #11202